arrow-json/src/reader/struct_array.rs

```diff
@@ -216,46 +238,90 @@ impl ArrayDecoder for StructArrayDecoder {
 
         let mut cur_idx = pos + 1;
         while cur_idx < end_idx {
-            // Read field name
             let field_name = match tape.get(cur_idx) {
                 TapeElement::String(s) => tape.get_string(s),
-                _ => return false,
+                _ => {
+                    return ErrorMarker::err(
+                        row_idx,
+                        cur_idx,
+                        FailureKind::TypeMismatch,
+                        Arc::clone(&self.data_type),
+                    );
+                }
             };
 
-            // Update child pos if match found
             match fields.iter().position(|x| x.name() == field_name) {
                 Some(field_idx) => {
                     let child_pos = cur_idx + 1;
-                    if !self.decoders[field_idx].validate_row(tape, child_pos) {
-                        return false;
+
+                    if let Err(mut child_errors) =
+                        self.decoders[field_idx].validate_row(tape, child_pos, row_idx)
+                    {
+                        // Add field name to child errors that don't have one
+                        for error in &mut child_errors {
+                            if error.field_name.is_none() {
+                                error.field_name = Some(field_name);
+                            }
+                        }
+                        return Err(child_errors);
                     }
+
                     validated_fields[field_idx] = true;
                 }
                 None => {
                     if self.strict_mode {
-                        return false;
+                        // Custom field_name - can't use helper
+                        return Err(vec![ErrorMarker {
```

Comment:
I'm a bit confused that we're both collecting errors (to be reported later) and also early returning on many errors, in which case we won't collect them.